core/cas-server-core-authentication-api/src/main/java/org/apereo/cas/authentication/ByCredentialTypeAuthenticationHandlerResolver.java

```diff
@@ -20,8 +23,20 @@ public ByCredentialTypeAuthenticationHandlerResolver(final Class<? extends Crede
         this(CollectionUtils.wrapSet(credentials));
     }
 
+    @Override
+    public Set<AuthenticationHandler> resolve(final Set<AuthenticationHandler> candidateHandlers, final AuthenticationTransaction transaction) {
+        final Set<Credential> supportedCreds = supported(transaction.getCredentials());
+        return candidateHandlers.stream().filter(h -> supportedCreds.stream().anyMatch(c -> h.supports(c)))
+                .collect(Collectors.toSet());
+    }
+
     @Override
     public boolean supports(final Set<AuthenticationHandler> handlers, final AuthenticationTransaction transaction) {
         return credentials.stream().anyMatch(transaction::hasCredentialOfType);
     }
+
+    private Set<Credential> supported(final Collection<? extends Credential> candidateCredentials) {
+        return candidateCredentials.stream().filter(c -> credentials.contains(c.getClass()))
+                .collect(Collectors.toSet());
```

Comment:
collect to to a set that can preserve order?

Reply:
Will do